clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -980,6 +1002,28 @@ void handleCoordinatorReady() {
                 null;
         this.coordinatorSupportsBumpingEpoch = initProducerIdVersion != null &&
                 initProducerIdVersion.maxVersion() >= 3;
+
+        if (nodeApiVersions == null) return;
+
+        if (nodeApiVersions.finalizedFeatures() != null) {
+            /*
+                To enable the transaction V2, it requires:
+                1. transaction.version finalized version >= 2
+                2. The ProduceRequest max version > ProducerRequest.LAST_BEFORE_TRANSACTION_V2_VERSION
```

Comment:
Yeah, if we enable v2 before all brokers support new produce requests, that would break the protocol, but this check doesn't do it correctly anyway because it's just a coordinator and we may produce to different brokers.